nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/main/java/org/nuxeo/ecm/platform/comment/impl/CommentManagerImpl.java

```diff
@@ -344,39 +346,53 @@ public DocumentModel createComment(DocumentModel docModel, DocumentModel parent,
+            Map<String, Object> ctxMap = Collections.singletonMap(ResourceAdapter.CORE_SESSION_CONTEXT_KEY,
+                    session);
+            RelationManager relationManager = Framework.getService(RelationManager.class);
+            Graph graph = relationManager.getGraph(config.graphName, session);
+            Resource commentResource = relationManager.getResource(config.commentNamespace, comment, ctxMap);
+            if (commentResource == null) {
+                throw new NuxeoException("Could not adapt document model to relation resource ; "
+                        + "check the service relation adapters configuration");
+            }
+            Resource predicate = new ResourceImpl(config.predicateNamespace);
+
+            List<Statement> statementList = graph.getStatements(commentResource, predicate, null);
+            if (graph instanceof JenaGraph) {
+                // XXX AT: BBB for when repository name was not included in the
+                // resource uri
+                Resource oldDocResource = new QNameResourceImpl(config.commentNamespace, comment.getId());
+                statementList.addAll(graph.getStatements(oldDocResource, predicate, null));
+            }
 
-        List<DocumentModel> docList = new ArrayList<>();
-        for (Statement stmt : statementList) {
-            QNameResourceImpl subject = (QNameResourceImpl) stmt.getObject();
-            DocumentModel docModel = (DocumentModel) relationManager.getResourceRepresentation(config.documentNamespace,
-                    subject, ctxMap);
-            if (docModel == null) {
-                log.warn("Could not adapt comment relation subject to a document "
-                        + "model; check the service relation adapters configuration");
-                continue;
+            List<DocumentModel> docList = new ArrayList<>();
+            for (Statement stmt : statementList) {
+                QNameResourceImpl subject = (QNameResourceImpl) stmt.getObject();
+                DocumentModel docModel = (DocumentModel) relationManager.getResourceRepresentation(config.documentNamespace,
+                        subject, ctxMap);
+                if (docModel == null) {
+                    log.warn("Could not adapt comment relation subject to a document "
+                            + "model; check the service relation adapters configuration");
+                    continue;
+                }
+                // detach the document as it was loaded by a system session, not the user session.
+                docModel.detach(true);
+                docList.add(docModel);
             }
-            docList.add(docModel);
-        }
-        return docList;
+            return docList;
+        });
+    }
 
+    /**
+     * @since 11.1
+     */
+    protected List<DocumentModel> doPrivileged(CoreSession session, String repositoryName,
+            Function<CoreSession, List<DocumentModel>> function) {
+        if (session == null) {
```

Comment:
Do we have legitimate cases where `session == null`? I don't see why we need this, the old code didn't seem to need it. If there really is a need, then please add a comment explaining the use case.

Reply:
The comments are often detached. Relying on the session can end up in an NPE like on this 8.10 failed test:
https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-nalkotob-8.10/13/testReport/org.nuxeo.box.api.test.comment/BoxCommentTest/itCanManageComment/

```
java.lang.NullPointerException
	at org.nuxeo.ecm.core.api.UnrestrictedSessionRunner.isUnrestricted(UnrestrictedSessionRunner.java:106)
	at org.nuxeo.ecm.core.api.UnrestrictedSessionRunner.<init>(UnrestrictedSessionRunner.java:61)
	at org.nuxeo.ecm.core.api.CoreInstance$2.<init>(CoreInstance.java:305)
	at org.nuxeo.ecm.core.api.CoreInstance.doPrivileged(CoreInstance.java:305)
	at org.nuxeo.ecm.platform.comment.impl.CommentManagerImpl.getDocumentsForComment(CommentManagerImpl.java:388)
```

So we rely on the session only if it's not null

Reply:
Ok